chore(cli): Add documentation for how to use the http api endpoints (#726)
ceramicnetwork/js-ceramic: docs-src/http-api.md

> The Ceramic HTTP API allows you to create and update documents on a remote Ceramic node. These endpoints allows you to manually make requests to a Ceramic node. This is also a good place to start if you want to build a Ceramic client in a new language. 

Reviewer: ```suggestion
The Ceramic HTTP API allows you to create, update, and query documents on a remote Ceramic node. These endpoints allows you to manually make requests to a Ceramic node. This is also a good place to start if you want to build a Ceramic client in a new language. 
```

        },
        "linkedBlock": "pGJpZNgqWCYAAYUBEiDRQJ7VCtGQWcLlmFpitGoSP35ntX7fKJeFWJ8zKz2+Z2RkYXRhgaNib3BjYWRkZHBhdGhlL21vcmVldmFsdWUY6mRwcmV22CpYJgABhQESINFAntUK0ZBZwuWYWmK0ahI/fme1ft8ol4VYnzMrPb5nZmhlYWRlcqFrY29udHJvbGxlcnOA"
      }
    }
  ]
}
```

## Apply a record to a document
*`⚠️ This method is not available if gateway mode is enabled.`*

In order to modify a document we apply a record to its docment log. This record usually contains a signature over a *json-patch* diff describing a modification to the document contents. The record also needs to contain pointers to the previous record and other metadata. You can read more about this in the [Ceramic Specification](https://github.com/ceramicnetwork/ceramic/blob/master/SPECIFICATION.md#document-records). Different document types may have different formats for their records. If you want to see an example implementation for how to construct a record you can have a look at the implementation of the [[TileDoctype]].

### Request
`POST /api/v0/records`

The request body should contain the following fields:

- `docId` - the name of the doctype to use, string

> - `record` - the genesis content of the document (will differ per doctype)

Reviewer: ```suggestion
- `record` - the content of the record to apply (will differ per doctype)
```

> - `docOpts` - options for the document creation [[DocOpts]]

Reviewer: ```suggestion
- `docOpts` - options for the document update [[DocOpts]] (optional)
```

The Ceramic HTTP API allows you to create and update documents on a remote Ceramic node. These endpoints allows you to manually make requests to a Ceramic node. This is also a good place to start if you want to build a Ceramic client in a new language. 

*`ℹ️ Not all of these methods are exposed if the node runs in gateway mode.`*

# Documents

> The documents endpoint is used to create new documents, load documents from their [[DocID]] or from their genesis content. 

Author: fixed.

## Get the state of a document
Load the state of a document given it's [[DocID]].

### Request
`GET /api/v0/documents/:docid`

Here, `:docid` should be replaced by the string representation of the DocID of the document that is being requested.

### Response

The response body contains the following fields:

- `docId` - the [[DocID]] of the requested document as string
- `state` - the state of the requested document as [[DocState]]

### Example

```bash
$ curl http://localhost:7007/api/v0/documents/kjzl6cwe1jw147r7878h32yazawcll6bxe5v92348cxitif6cota91qp68grbhm

{
  "docId": "kjzl6cwe1jw147r7878h32yazawcll6bxe5v92348cxitif6cota91qp68grbhm",
  "state": {
    "doctype": "tile",
    "content": {
      "Ceramic": "pottery"
    },
    "metadata": {
      "schema": null,
      "controllers": [
        "did:key:z6MkfZ6S4NVVTEuts8o5xFzRMR8eC6Y1bngoBQNnXiCvhH8H"
      ]
    },
    "signature": 2,
    "anchorStatus": "PENDING",
    "log": [{
      "cid": "bagcqceramof2xi7kh6qblirzkbc7yulcjcticlcob6uvdrx3bexgks37ilva",
      "type": 0
    }],
    "anchorScheduledFor": "12/15/2020, 2:45:00 PM"
  }
}
```

## Create a document from genesis record

> Create a new document, or load a document from it's genesis content (these are equivalent actions). The genesis content may be signed (a DagJWS for the `tile` doctype), or unsigned in some cases.

Author: > If we had a way to get a CID from some content without actually publishing that content to IPFS

This is possible (technically), don't see how that changes this statement though?

> Create a new document, or load a document from it's genesis content (these are equivalent actions). The genesis content may be signed (a DagJWS for the `tile` doctype), or unsigned in some cases.

Reviewer: Then they wouldn't be equivalent actions anymore.  It would be possible to load a document from its genesis contents without creating the document if it doesn't exist

        "type": 0
      }
    ],
    "anchorScheduledFor": "12/30/2020, 1:45:00 PM"
  },
  "kjzl6cwe1jw149rledowj0zi0icd7epi9y1m5tx4pardt1w6dzcxvr6bohi8ejc": {
    "doctype": "tile",
    "content": {
      "Document": "A"
    },
    "metadata": {
      "schema": null,
      "controllers": [
        "did:key:z6MkfZ6S4NVVTEuts8o5xFzRMR8eC6Y1bngoBQNnXiCvhH8H"
      ]
    },
    "signature": 2,
    "anchorStatus": "FAILED",
    "log": [
      {
        "cid": "bagcqcerawq5h7otlkdwuai7vhogqhs2aeaauwbu2aqclrh4iyu5h54qqogma",
        "type": 0
      }
    ],
    "anchorScheduledFor": "12/30/2020, 1:45:00 PM"
  },
  "kjzl6cwe1jw147w3xz3xrcd37chh2rz4dfra3imtnsni385rfyqa3hbx42qwal0": {
    "doctype": "tile",
    "content": {
      "Document": "B"
    },
    "metadata": {
      "schema": null,
      "controllers": [
        "did:key:z6MkfZ6S4NVVTEuts8o5xFzRMR8eC6Y1bngoBQNnXiCvhH8H"
      ]
    },
    "signature": 2,
    "anchorStatus": "FAILED",
    "log": [
      {
        "cid": "bagcqceranecdjzw4xheudgkr2amjkntpktci2xv44d7v4hbft3ndpptid6ka",
        "type": 0
      }
    ],
    "anchorScheduledFor": "12/30/2020, 1:45:00 PM"
  }
}
```

# Records

> The records endpoint provides lower level access to the data structure of a Ceramic document. It is also the enpoint that has to be used in order to update a document, by adding a new record.

Reviewer: a nitpick, but saying "has to" makes it sound like a negative.  "is" just indicates that this is the reality, without any judgement associated

      }
    },
    {
      "cid": "bagcqcera3fkje7je4lvctkam4fvi675avtcuqgrv7dn6aoqljd5lebpl7rfq",
      "value": {
        "jws": {
          "payload": "AXESINm6lI30m3j5H2ausx-ulXj-L9CmFlOTZBZvJ2O734Zt",
          "signatures": [
            {
              "signature": "zsLJbBSU5xZTQkYlXwEH9xj_t_8frvSFCYs0SlVMPXOnw8zOJOsKnJDQlUOvPJxjt8Bdc_7xoBdmcRG1J1tpCw",
              "protected": "eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDprZXk6ejZNa2ZaNlM0TlZWVEV1dHM4bzV4RnpSTVI4ZUM2WTFibmdvQlFOblhpQ3ZoSDhII3o2TWtmWjZTNE5WVlRFdXRzOG81eEZ6Uk1SOGVDNlkxYm5nb0JRTm5YaUN2aEg4SCJ9"
            }
          ],
          "link": "bafyreigzxkki35e3pd4r6zvowmp25fly7yx5bjqwkojwiftpe5r3xx4gnu"
        },
        "linkedBlock": "pGJpZNgqWCYAAYUBEiDRQJ7VCtGQWcLlmFpitGoSP35ntX7fKJeFWJ8zKz2+Z2RkYXRhgaNib3BjYWRkZHBhdGhlL21vcmVldmFsdWUY6mRwcmV22CpYJgABhQESINFAntUK0ZBZwuWYWmK0ahI/fme1ft8ol4VYnzMrPb5nZmhlYWRlcqFrY29udHJvbGxlcnOA"
      }
    }
  ]
}
```

## Apply a record to a document
*`⚠️ This method is not available if gateway mode is enabled.`*

> In order to modify a document we apply a record to its docment log. This record usually contains a signature over the *json-patch* of the document contents. The record also needs to contain pointers to the previous record and other metadata. You can read more about this in the [Ceramic Specification](https://github.com/ceramicnetwork/ceramic/blob/master/SPECIFICATION.md#document-records). Different document types may have different formats for their records. If you want to see an example implementation for how to construct a record you can have a look at the implementation of the [[TileDoctype]].

Reviewer: ```suggestion
In order to modify a document we apply a record to its docment log. This record usually contains a signature over a *json-patch* diff describing a modification to the document contents. The record also needs to contain pointers to the previous record and other metadata. You can read more about this in the [Ceramic Specification](https://github.com/ceramicnetwork/ceramic/blob/master/SPECIFICATION.md#document-records). Different document types may have different formats for their records. If you want to see an example implementation for how to construct a record you can have a look at the implementation of the [[TileDoctype]].
```

### Request
`POST /api/v0/records`

> Here, `:docid` should be replaced by the string representation of the DocID of the document that is being requested.

Reviewer: This whole section looks like it was copy and pasted from elsewhere without being updated

    "content": {
      "title": "My first Document"
    },
    "metadata": {
      "schema": null,
      "controllers": [
        "did:key:z6MkfZ6S4NVVTEuts8o5xFzRMR8eC6Y1bngoBQNnXiCvhH8H"
      ]
    },
    "signature": 2,
    "anchorStatus": "PENDING",
    "log": [
      {
        "cid": "bagcqcera2faj5vik2giftqxftbngfndkci7x4z5vp3psrf4flcptgkz5xztq",
        "type": 0
      },
      {
        "cid": "bagcqcera3fkje7je4lvctkam4fvi675avtcuqgrv7dn6aoqljd5lebpl7rfq",
        "type": 1
      }
    ],
    "anchorScheduledFor": "12/30/2020, 1:15:00 PM",
    "next": {
      "content": {
        "title": "My first Document",
        "more": 234
      },
      "metadata": {
        "schema": null,
        "controllers": []
      }
    }
  }
}

```

# Pins

The pins api endpoint can be used to manipulate the pin set. The pin set is all of the documents that a node maintains the state of. Any document opened by the node that is not pinned will eventually be garbage collected from the node.

## Add a document to the pin set
This method adds the document with the given [[DocID]] to the pin set. 

*`⚠️ This method is not available if gateway mode is enabled.`*

### Request

> `POST /api/v0/pins/:docid`

Reviewer: huh, this does indeed match what we have in code, but is surprising to me.  Since this is a POST request not a GET, shouldn't the docid be encoded as an arg to the `pins` endpoint, rather than encoded in the URL?

CC @simonovic86

Here, `:docid` should be replaced by the string representation of the DocID of the document that is being requested.

### Response

* `docId` - the [[DocID]] of the document which was pinned, string

> * `isPinned` - whether the document was pinned, boolean - `true`

Reviewer: another comment less about the docs than about the actual impl, but isn't this field redundant and unnecessary?  It can never come back as `false` currently, because if there's a problem we return a non-ok http status response.

@simonovic86

Author: Yeah, Agree that it could be removed.

Reviewer: let's just remove the mention of `isPinned` from the docs so we can remove it later.  We should also add a sentence to clarify that an OK http response indicates the pin/unpin operation was successful

### Example

```bash
curl http://localhost:7007/api/v0/pins/k2t6wyfsu4pg2qvoorchoj23e8hf3eiis4w7bucllxkmlk91sjgluuag5syphl -X POST

{
  "docId": "k2t6wyfsu4pg2qvoorchoj23e8hf3eiis4w7bucllxkmlk91sjgluuag5syphl",
  "isPinned": true
}
```

## Remove a document from the pin set
This method removes the document with the given [[DocID]] from the pin set.

*`⚠️ This method is not available if gateway mode is enabled.`*

### Request

> `DELETE /api/v0/pins/:docid`

Reviewer: same comment as above about encoding the docid in the request vs in the url

Here, `:docid` should be replaced by the string representation of the DocID of the document that is being requested.

### Response

* `docId` - the [[DocID]] of the document which was unpinned, string

> * `isPinned` - whether the document was pinned, boolean - `false`

Reviewer: this doesn't actually return whether the document was pinned previously.  It is just always `false` if the unpin command was successful

### Example

```bash
curl http://localhost:7007/api/v0/pins/k2t6wyfsu4pg2qvoorchoj23e8hf3eiis4w7bucllxkmlk91sjgluuag5syphl -X DELETE

{
  "docId": "k2t6wyfsu4pg2qvoorchoj23e8hf3eiis4w7bucllxkmlk91sjgluuag5syphl",
  "isPinned": false
}
```

> ## Check if a document is in the pin set

Reviewer: ```suggestion
## Check if a specific document is in the pin set
```

This method is used to check if a particular document is in the pin set.

> ### Request

Reviewer: a nitpick, but I find the flow of this document easier to read if the "list all documents in the pin set" section comes before the "check if document is in the pin set" section

Author: good point.

`GET /api/v0/pins/:docid`

Here, `:docid` should be replaced by the string representation of the DocID of the document that is being requested.

### Response

> * `pinnedDocIds` - an array of [[DocID]] strings that are in the pin set

Reviewer: ```suggestion
* `pinnedDocIds` - an array containing the specified [[DocID]] string if that document is pinned, or an empty array if that document is not pinned
```
